src/Umbraco.Web/Editors/CodeFileController.cs

```diff
@@ -0,0 +1,155 @@
+﻿using AutoMapper;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web.Http;
+using Umbraco.Core.Models;
+using Umbraco.Web.Models.ContentEditing;
+using Umbraco.Web.Mvc;
+using Umbraco.Web.WebApi;
+
+namespace Umbraco.Web.Editors
+{
+    [PluginController("UmbracoApi")]
+    public class CodeFileController : UmbracoAuthorizedJsonController
+    {
+
+        public HttpResponseMessage PostCreate(string type, CodeFileDisplay display)
```

Comment:
There is no validation being done on anything here, all validation that is done client side also must be done server side. See: http://issues.umbraco.org/issue/U4-9324#comment=67-33788 (yes, we need to properly document this). Your model needs data annotations added to it then you could easily 'just' add `[ValidationFilter]` to this action assuming that further server side validation doesn't need to take place - but it probably does since you are sending up virtual paths which may map to illegal server side paths which could lead to malicious activity.